research/cifar10/ditto/client.py

```diff
@@ -0,0 +1,186 @@
+import argparse
+import os
+from logging import INFO
+from pathlib import Path
+from typing import Dict, Optional, Sequence, Tuple
+
+import flwr as fl
+import torch
+import torch.nn as nn
+from flwr.common.logger import log
+from flwr.common.typing import Config
+from torch.nn.modules.loss import _Loss
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
+
+from fl4health.checkpointing.checkpointer import BestLossTorchCheckpointer
+from fl4health.checkpointing.client_module import ClientCheckpointModule
+from fl4health.clients.ditto_client import DittoClient
+from fl4health.utils.load_data import load_cifar10_data, load_cifar10_test_data
+from fl4health.utils.losses import LossMeterType
+from fl4health.utils.metrics import Accuracy, Metric
+from fl4health.utils.random import set_all_random_seeds
+from fl4health.utils.sampler import DirichletLabelBasedSampler
+from research.cifar10.model import ConvNet2
+
+NUM_CLIENTS = 10
+
+
+class CifarDittoClient(DittoClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        client_number: int,
+        learning_rate: float,
+        lam: float,
+        heterogeneity_level: float,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        checkpointer: Optional[ClientCheckpointModule] = None,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+            lam=lam,
+        )
+        self.client_number = client_number
+        self.heterogeneity_level = heterogeneity_level
+        self.learning_rate: float = learning_rate
+
+        assert 0 <= client_number < NUM_CLIENTS
+        log(INFO, f"Client Name: {self.client_name}, Client Number: {self.client_number}")
+
+    def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
+        batch_size = self.narrow_config_type(config, "batch_size", int)
+        n_clients = self.narrow_config_type(config, "n_clients", int)
+        # Set client-specific hash_key for sampler to ensure heterogneous data distribution among clients
+        sampler = DirichletLabelBasedSampler(
+            list(range(10)),
+            sample_percentage=1.0 / n_clients,
```

Comment:
Yeah that is true, but I think we talked about this and decided that it should not be a deal breaker to have overlapping samples among clients. It might be though. To make sure I think, it worths trying this sampling strategy that you mentioned.

Reply:
Given this new dirichlet sampling implementation, do you think I still need to add this `load_cifar10_partitioned_data` ?

Reply:
No I think we're good to just move to the Dirichlet partitioner for this 🙂